### ISSUE TITLE
Parse error when pattern synonyms are imported

### ISSUE THREAD

**Issue by eigengrau** [haskell-suite/haskell-src-exts#262]
It seems that even with #197 merged, haskell-src-exts will fail parsing a file when pattern synonyms are imported.

E. g., the following:

```
import Foo.Bar (pattern SomePattern)
```

Still causes stylish-haskell to fail with:

```
Language.Haskell.Stylish.Parse.parseModule: could not parse src/YampaGame/Main.hs: ParseFailed (SrcLoc "<unknown>.hs" 29 45) "Parse error: pattern"
```

**Comment by pjonsson** [comment]
@mpickering: will you add the support for pattern synonym imports?

**Comment by mpickering** [comment]
I had a quick look and it seems like my original implementation of pattern exports might be a bit confused as well. 

https://hackage.haskell.org/package/haskell-src-exts-1.16.0.1/docs/Language-Haskell-Exts-Annotated-Syntax.html#t:ExportSpec

In the implementation I followed the example from `ExplicitNamespaces` and parsed it as an `EVar PatternNamespace` but looking now more closely at the documentation it seems that "a class or datatype exported abstractly, or a type synonym." should instead be `EAbs` or an entirely new constructor. It seems there is this same confusion for `ExplicitNamespaces`, maybe the proper implementation is to actually add a new constructor for these separate types of exports?

With the current implementation the following file is parsed unpredictably (imo). 

```
{-# LANGUAGE ExplicitNamespaces #-}
module Foo (Syn, type Syn) where

type Syn = Int
```

```
ParseOk
  (Module
     (SrcLoc "test.hs" 1 1)
     (ModuleName "Foo")
     [ LanguagePragma
         (SrcLoc "test.hs" 1 1) [ Ident "ExplicitNamespaces" ]
     ]
     Nothing
     (Just
        [ EAbs (UnQual (Ident "Syn"))
        , EVar TypeNamespace (UnQual (Ident "Syn"))
        ])
     []
     [ TypeDecl
         (SrcLoc "test.hs" 4 1)
         (Ident "Syn")
         []
         (TyCon (UnQual (Ident "Int")))
     ])
```

The first `Syn` is parsed as `EAbs` but by prefixing the `type` keyword it gets parsed as `EVar`. GHC parses `type` and `pattern` exports to `IEThingAbs` which I think it similar to `EAbs`. 

What do you think the best plan of action is?

**Comment by pjonsson** [comment]
Ok, this turned out to be trickier than I thought so I'm not sure what the right design of the AST is. I agree that the namespace on `EVar` seems misplaced. Do you have a name suggestion for the new constructor?

Since there isn't a clear right course of action I suggest we sleep on it so we don't make things worse than they currently are. We should fix the pretty printing of the contexts in the pattern type signatures so that our test suite can catch regressions before we start refactoring those parts of the AST.

**Comment by pjonsson** [comment]
Ticket #93 is related to the rationale behind the current design of ExplicitNamespaces.

**Comment by mpickering** [comment]
I think the right thing to do is move the field from `EVar` to `EAbs` at least in the case of `TypeNamespace`.

After all it says "a class or datatype exported abstractly, or a type synonym." with `ExplicitNamespaces` you are exporting a type whether from a type family or a type synonym or a normal datatype. 

As for `PatternSynonyms` you are exporting a *pattern* and  potentially a *constructor* which ordinarily you can't do without exporting the type as well using `EThingAll` or `EThingWith` so there's not the same correspondance as with `ExplicitNamespaces`. For that reason I think it would be best to have a separate constructor which corresponds to pattern exports.

**Comment by pjonsson** [comment]
The problem with the first implementation of explicit namespaces was that it admitted many weird ASTs. Roman is right in the other ticket when he's pointing out things that don't make sense. We should use an example that contains potential ambiguity with explicit namespaces though so let's borrow the one from the GHC manual in section 7.4.4: "There is now some potential ambiguity in import and export lists; for example if you write `import M( (+) )` do you mean the function `(+)` or the type constructor `(+)`?". For better or worse, the current implementation of explicit namespaces parses both the function and the type constructor as an `EVar`.

I think looking at export lists will lead us astray because the entire module is right in front of us and we might not manage to ignore what we see. Let's consider import lists for our examples instead because we can't peek at the implementation there.

We have the same potential ambiguity in import lists between patterns and classes/datatypes, does `import M( Syn )` mean the pattern `Syn` or the class `Syn`? Suppose for arguments sake that we invent `data Namespace2 = NoNamespace | PatternNamespace`. We could chuck that `Namespace2` into `EAbs` just like we do with `Namespace` on `EVar`. It's not very pretty but the advantage would be that the same (twisted (parser)) logic applied to `EVar` and `EAbs`.

I'm not suggesting we do this but if you ignore the beauty-aspects, would the `Namespace2` approach be sufficient to support parsing of pattern synonyms for imports and exports?

**Comment by mpickering** [comment]
Adding an extra field to `EVar` would be sufficient for parsing for parsing pattern exports and imports as they both have a field for `QName` so this is purely a semantic distinction we have to make. The original implementation does look wrong but could have been salvaged I think by changing `ImportSpec` to `QName` rather than mashing it in with `EVar`. 

Is there a strong argument for *not* having a separate constructor for both of them? `EType` and `EPattern` which each have ` QName` field? That seems the best solution?

**Comment by pjonsson** [comment]
Having many isomorphic constructors in the same datatype is usually a sign of failing to capture some common structure in a sensible way. We suffer from that in our prettyprinter which is hairy to say the least and probably full of bugs that have been fixed in one place but not in another place. Our suffering is not as important as our users though--if they want to treat things uniformly they have to pattern match on a bunch of constructors and do the same thing for all of them. There is a fine balance--if we generalize things too much we will sacrifice type safety in our AST and that will also cause suffering for ourselves and our users.

It seems clear that `EThingAll`, `EThingWith`, and `EModuleContents` are always in `NoNamespace` so they should not have a namespace modifier.

That leaves us with `EVar` and `EAbs`. Ignoring the namespace part, they can represent the same things since they both carry a `QName`.

If we add a namespace to `EAbs` they would be identical but `EVar` would be used to represent things in the type namespace and `EAbs` would be used to represent things in the pattern namespace. Not a good design.

If we move the namespace from `EVar` to `EAbs` your example with `Syn` would make more sense. A similar example with `(+)` could be constructed and it makes sense that the parse tree is different for the type and the function. We can use `PatternNamespace` that we currently have in namespace to handle  patterns with `EAbs` as well. We also need to update the documentation for `EAbs` so that it's general enough to be any abstraction. This is clearly better than the alternative of having the namespace in both constructors and only use disjoint subsets of the constructors in each. Patterns as explained by the GHC manual do abstract the underlying implementation so I'm not sure this is a bad fit at the conceptual level. The example of importing the Just constructor is also an abstraction that happens to coincide with the concrete implementation.

A third option is to remove the patterns from the Namespace datatype. We need a new constructor or some other way to handle patterns with this alternative.

In your initial comment you didn't seem to have any strong opinion on the GHC design with having patterns in `EAbs` and 2 comments back you seem to have gravitated towards having a separate constructor. What was your reasoning against `EAbs` and for the separate constructor?

**Comment by mpickering** [comment]
I think changing `EVar` and modifying `EAbs` is the right thing to do with your comments above.

**Comment by pjonsson** [comment]
Great! I added some tickets to track things. They should be merged in order:

1. #264
1. #265 
1. #262

I can do #265 if you don't have time. Can you please fix #264 and submit a pull request though? Adding "Fixes #ticket" to the commit message closes the ticket automatically when it's merged to HEAD.

**Comment by pjonsson** [comment]
Fixed with d3badcc3ab12e868483686a8969e2fb0567de589.

**Comment by eigengrau** [comment]
Working nicely, thanks!
